src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -339,19 +342,26 @@ int ValidateAndGetLengthGeneric(ICollection<TElement> value, ref NpgsqlLengthCac
             4 +              // has_nulls (unused)
             4 +              // type OID
             1 * 8 +          // number of dimensions (1) * (length + lower bound)
-            4 * value.Count; // sum of element lengths
+            4 * readOnlyMemory.Length; // sum of element lengths
 
         lengthCache.Set(0);
         var elemLengthCache = lengthCache;
 
-        foreach (var element in value)
+        if (typeof(TElement).IsValueType)
```

Comment:
I'm pretty sure this type of optimization is wrong: a good example is nullable types, which can be null. Another example is `TimestampTzHandler`, where we check a `DateTime`'s kind.